presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java

```diff
@@ -330,14 +338,59 @@ private boolean checkTablePermission(ConnectorSecurityContext context, SchemaTab
         return false;
     }
 
+    private boolean checkTableAnyPermission(ConnectorSecurityContext context, SchemaTableName tableName)
+    {
+        if (INFORMATION_SCHEMA_NAME.equals(tableName.getSchemaName())) {
+            return true;
+        }
+        if (!canAccessSchema(context, tableName.getSchemaName(), AccessMode.READ_ONLY)) {
+            return false;
+        }
+        for (TableAccessControlRule rule : tableRules) {
+            Optional<Set<TablePrivilege>> tablePrivileges = rule.match(context.getIdentity().getUser(), tableName);
+            if (tablePrivileges.isPresent()) {
+                if (tablePrivileges.get().isEmpty()) {
+                    return false;
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
     private boolean isSchemaOwner(ConnectorSecurityContext context, String schemaName)
     {
         for (SchemaAccessControlRule rule : schemaRules) {
-            Optional<Boolean> owner = rule.match(context.getIdentity().getUser(), schemaName);
+            Optional<Boolean> owner = rule.isSchemaOwner(context.getIdentity().getUser(), schemaName);
             if (owner.isPresent()) {
                 return owner.get();
             }
         }
         return false;
     }
+
+    private boolean canAccessSchema(ConnectorSecurityContext context, String schemaName, AccessMode requiredAccess)
+    {
+        for (SchemaAccessControlRule rule : schemaRules) {
+            Optional<AccessMode> accessMode = rule.match(context.getIdentity().getUser(), schemaName);
+            if (accessMode.isPresent()) {
+                return accessMode.get().implies(requiredAccess);
+            }
+        }
+        // if not rule match, we assume schema as READ_ONLY by default
+        return true;
+    }
+
+    private static AccessMode calculateAccessModeFromTablePrivileges(TablePrivilege... tablePrivileges)
+    {
+        AccessMode accessMode = AccessMode.NONE;
+        for (TablePrivilege tablePrivilege : tablePrivileges) {
+            if (tablePrivilege.equals(SELECT) && accessMode.equals(AccessMode.NONE)) {
+                accessMode = AccessMode.READ_ONLY;
+                continue;
+            }
+            accessMode = AccessMode.ALL;
```

Comment:
Table privilege should not imply we have ALL schema access.

It should be possible to have read-only access to a schema and whatever access (full / OWNERSHIP / GRANT_SELECT / ...) to a table within that schema.

Reply:
you mean we can not derive schema access mode just base on required privilege of one table, right ?

Reply:
I renamed ` calculateAccessModeFromTablePrivileges ` to ` calculateRequiredSchemaAccessModeFromTablePrivileges `

Reply:
I still don't understand why this `calculateRequiredSchemaAccessModeFromTablePrivileges` ("required schema access for given table privilege/access") is needed. I would assume such method would always return READ_ONLY (and thus wouldn't be needed).

@kokosing can you help me?